### PR TITLE
[FRONTEND] OpenAI `tools` support named functions

### LINKED ISSUE #1869: Support `tools` and `tool_choice` parameter in OpenAI compatible service
Also aliased as `functions` and `function_call` in deprecated parameters.

https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools

After #1756 is merged (thanks @Tostino!), it should be straightforward to add this as a core parameter to OpenAI compatible service. This will help unlock client libraries using similar interface. Do note that the underlying model need to support function calling (e.g. OpenHermes) and prompt engineering might be needed. 

Also see @dongxiaolong's example here: https://github.com/vllm-project/vllm/pull/1756#issuecomment-1827064922

### PR DESCRIPTION
**Pull request by br3no**
PARTIAL FIX #1869 (and #5008) -- only named tools

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by br3no** [comment]
@simon-mo 

This PR implements only the "named tool" feature of the OpenAI API, as sketched in my comment here: https://github.com/vllm-project/vllm/issues/1869#issuecomment-2124776890

This exposes the `guided_json` functionality in an OpenAI compatible way.

**Comment by simon-mo** [comment]
Thank you for this PR. Will review and merge once the logprobs is in. Let's get this in.

**Comment by K-Mistele** [comment]
#4656 is a more complete implementation of this that supports auto tool choice (for models that support it); just need to fix jinja templating for the tool usage system prompt since different models have different system prompt formats (e.g. mistral 7B instruct v0.3, vs. Nous Hermes' Hermes 2 Pro Llama 3 8B)

**Comment by br3no** [comment]
@K-Mistele, I'm aware of #4656. I believe that a step-by-step approach will more easily find it's way into vLLM, so starting with the smallest possible increment is a much easier and more manageable task.

Once this is merged, #4656 can continue with the remaining missing features.

**Comment by br3no** [comment]
@simon-mo this is ready for review now.

**Comment by br3no** [comment]
@simon-mo the failing tests seem to be infrastructure-related and unrelated with the change. I hope you have time to review this soon. I’d love to have this in the next release. Let me know if you have any feedback so I can address all issues before you cut the next tag!

File: tests/entrypoints/test_openai_server.py [hunk 1]
```diff
@@ -906,6 +906,194 @@ async def test_guided_choice_chat_logprobs(server, client: openai.AsyncOpenAI,
...
+            tools=[{
+                "type": "function",
+                "function": {
+                    "name": "dummy_function_name",
+                    "description": "This is a dummy function",
+                    "parameters": TEST_SCHEMA
+                }
+            }],
+            tool_choice="required")
+        ...
```
Comment (simon-mo) [inline tests/entrypoints/test_openai_server.py:1029]:
```suggestion
```

File: tests/entrypoints/test_openai_server.py [hunk 2]
```diff
@@ -906,6 +906,194 @@ async def test_guided_choice_chat_logprobs(server, client: openai.AsyncOpenAI,
...
+                                             messages=messages,
+                                             max_tokens=1000,
+                                             tool_choice={
+                                                 "type": "function",
+                                                 "function": {
+                                                     "name":
+                                                     "dummy_function_name"
+                                                 }
+                                             })
+        ...
```
Comment (simon-mo) [inline tests/entrypoints/test_openai_server.py:1073]:
```suggestion
```

File: tests/entrypoints/test_openai_server.py [hunk 3]
```diff
@@ -906,6 +906,194 @@ async def test_guided_choice_chat_logprobs(server, client: openai.AsyncOpenAI,
...
+                    "parameters": TEST_SCHEMA
+                }
+            }],
+            tool_choice={
+                "type": "function",
+                "function": {
+                    "name": "nondefined_function_name"
+                }
+            })
+        ...
```
Comment (simon-mo) [inline tests/entrypoints/test_openai_server.py:1094]:
```suggestion
```

File: vllm/model_executor/guided_decoding/__init__.py [hunk 4]
```diff
@@ -23,3 +26,26 @@ async def get_guided_decoding_logits_processor(
...
+    # the legacy completion API does not support tool use
+    if type(request) == CompletionRequest:
+        return request
+
+    # user has chosen to not use any tool
+    if request.tool_choice == "none":
+        return request
+
+    # user has chosen to use a named tool
+    if type(request.tool_choice) == ChatCompletionNamedToolChoiceParam:
```
Comment (simon-mo) [inline vllm/model_executor/guided_decoding/__init__.py:42]:
```suggestion
    if type(request.tool_choice) is ChatCompletionNamedToolChoiceParam:
```
`is` used everywhere else except here i think

**Review by simon-mo — APPROVED** [review]
This is great and thank you for breaking up the big PR. Most comments are cosmetic. Just to confirm, we are not appending the schema to the prompt and expecting users to do it? Should we add it in?

**Comment by br3no** [comment]
@simon-mo cosmetic changes pushed.

Yes, correct, we are not appending the schema to the prompt. 

This question will become more relevant once we add support for `auto` and `required`. Since this is highly dependent on the model and the use-case (e.g. language of the prompt), we will need to add support for configurable tool-prompt templates (see #4656, where @K-Mistele proposed just that, which I think is a good idea).

I think it's fair to have the user take care of this in this first iteration.

What do you think? I could add a documentation on the topic to make sure people know how to use this.

**Comment by br3no** [comment]
I added a suggestion for the documentation. Let me know what you think.

**Review by simon-mo — APPROVED** [review]

File: vllm/entrypoints/openai/serving_chat.py [hunk 5]
```diff
@@ -381,9 +395,22 @@ async def chat_completion_full_generator(
...
 
+            if request.tool_choice and type(
+                    request.tool_choice) is ChatCompletionNamedToolChoiceParam:
+                message = ChatMessage(
+                    role=role,
+                    content="",
+                    tool_calls=[
+                        ToolCall(function=FunctionCall(
+                            name=request.tool_choice.function.name,
+                            arguments=output.text))
```
Comment (aw632) [inline vllm/entrypoints/openai/serving_chat.py:406]: Question: how does this part work? It looks like it's just taking the model's output and passing that in as the arguments, but how are we guarantee that the output matches the argument's requirements?

Comment (br3no) [reply vllm/entrypoints/openai/serving_chat.py:406]: Outlines or lm-format-enforcer guarantee that for us.
